[Proposal] Maintain a stateless CSI implementation

ShyamsundarR opened the pull request:
Introduce CSI plugin configuration, per Ceph cluster that is
used for provisioning storage, in lieu of the existing config
map based volume store.

Signed-off-by: ShyamsundarR <srangana@redhat.com>

ShyamsundarR commented:
Related to discussions around #135, #205 and #206 

Requesting review comments on approach: @gman0 @travisn @rootfs @humblec @Madhu-1 

Tagging folks who helped brainstorm for a review as well @phlogistonjohn @raghavendra-talur @JohnStrunk

JohnStrunk commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+- We cannot, deal with multiple Ceph clusters backed by a single CSI plugin
+  instance, due to the strict 1:1 mapping between the same, and thus do any
+  form of intelligent provisioning across Ceph clusters for a request
+- **Needs Investigation** Does this hinder Topology based volume provisioning?
```
Is a ceph cluster per topology how we would advocate deployment or would the ceph cluster span all topologies and use the crush map to direct allocations?

ShyamsundarR replied:
I really do not have an answer to the question. I have made references to the same in @liewegas comment as well. Once I learn more about what Ceph can do, I can attempt to answer the question.

In my **assumption**, a Ceph cluster is per Topology.

liewegas commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+  form of intelligent provisioning across Ceph clusters for a request
+- **Needs Investigation** Does this hinder Topology based volume provisioning?
+  As each instance is stuck to single Ceph cluster, we cannot provision from
+  different Topologies (each with their own Ceph clusters)
```
What is a Topology and how does it relate to a StorageClass?  Each StorageClass maps to a particular pool or fs in a particular ceph cluster, right?

ShyamsundarR replied:
This surely needs more investigation from my end, so answers maybe sketchy.

First up, IMO even the CSI k8s implementation is sketchy in this regard (it is alpha). ([reference](https://kubernetes-csi.github.io/docs/topology.html)).

Interesting point to note is point (2) in the above link that reads,
```
Ability for Kubernetes (users or components) to influence where a volume is
provisioned (e.g. provision new volume in either "zone 1" or "zone 2").
```

I read the above as, a request may come in with a Topology requirement. If the StorageClass already has the cluster and pool specified, the StorageClass is tied to a Topology in my understanding and if so we cannot decide where to provision from, it is decided by the StorageClass already. Which means we lose the ability to provision from a specific Topology, and each pod has to use a StorageClass that ties it to a Topology.

Unless, as @JohnStrunk asks, and I have questions as well around the same, can a Ceph cluster span so called Typologies? IOW, let's call them data centers (DC) in a federated cluster and can we have pools on DC1 and different pools in DC2, but all part of the same Ceph cluster?
- If yes, then specifying the cluster in the StorageClass would do, the request would come in with a Topology requirement, and pool selection should be done based on the same
- If no, then CSI driver would have to choose the cluster itself to provision from, and hence even that information should not come from the StorageClass

All this is speculation at best, currently the StorageClass **will** specify the cluster and pool that the volume should be provisioned from. In the future, based on the above proposal and based on better understanding of what Topology really entails, we can omit the same and add the logic within the CSI driver if needed.

ShyamsundarR commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+- Allow the CS and NS namespaces access to the above config maps and secrets,
+  in addition to the namespace that is responsible for creating and
+  maintaining it
+  - **Needs Investigation** Unsure if secrets and config maps can be
```
Good to know. Thank you.

The above comment covers where we may need RBAC and where not.

ShyamsundarR commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+    needed if we need operator managed Ceph instances to be able to pass
+    cluster configuration information to the CSI instance, and are in
+    different namespaces for e.g.
+- Details about the Ceph cluster are maintained in
```
We could for sure.

The caveat though is, if only certain pools are exposed to the CO system, then list operations will attempt to crawl/list images from all pools, which would be wasteful. Stating the pools that the current CO system would use, may help in this regard.

ShyamsundarR replied:
Current configuration only maintains the MONs as the storage class carries the pool/fsname details. Thus this is now pared down to the minimum for the requirement.

ShyamsundarR commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+  for the cluster and its secrets to act on the request
+  - **NOTE:** The cluster-id and pool information is specified in the
+    StorageClass for example
+  - **NOTE:** In the **future**, if it is desired that the CSI plugin CS
```
The CS does not have any opinion on this, which is correct.

Specifying the cluster and pool from the storage class again will work just fine, and initial implementation would actully expect the same.

The proposal leaves the ability to not do so, and let the CSI driver choose which cluster and pool to use for the requested provisioning, if this information is missing. My reasons are two fold, but not **substantiated** enough,
- Simplify the StorageClass in such a manner, that just like asking for gp2 instances from AWS, the StorageClass asks for image type X from Ceph, and does not bother with where it comes from. Makes application pod requesting storage from a more generically named StorageClass possible.
- Topology and how that impacts where the storage comes from

So, just to reiterate, this is just a possible **future** direction that is enabled, how we want to reach there and if it is a reality, still needs data to conclude on the strategy.

ShyamsundarR replied:
This is removed from the text. Retaining pool/fsname and clusterid to be specified in the storage class.

Madhu-1 commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+- Also, does not take away the capability, where needed, to use multiple
+  Ceph-CSI plugin instances (specialized using provisioner name) per CO
+  instance
+- Leaves the flexibility for the StorageClass to still override secrets, if such
```
for volume/snapshot creation  we need to verify that the resource exists and some other information like (size, does this volume has snapshot feature enabled, does this snapshot ID exists for other volumes) are we going call the CLI command to list/info the volume (parse the CLI response, based on the CLI response  take some action)
same for deletion also.

we can make use of setting user metadata on snapshot about parent volume and snapshot creation time.

ShyamsundarR replied:
> for volume/snapshot creation we need to verify that the resource exists and some other information like (size, does this volume has snapshot feature enabled, does this snapshot ID exists for other volumes) are we going call the CLI command to list/info the volume (parse the CLI response, based on the CLI response take some action)
> same for deletion also.

Yes, as we store no information we would be calling the CLI for all operations. I see snapshot creation as follows (against rbd),
- On Ceph we need to execute: `rbd snap create {pool-name}/{image-name}@{snap-name}`
- The above **needs** pool-name, image-name, snap-name
- The **inputs** as per the CSI spec are, source_volume_id, csi-snap-name
- From *source_volume_id*, extract *pool_name* and generate *ImageName*
- From *csi-snap-name*, generate *snap-ImageName*
- Check *snap-ImageName* existence on given *ImageName*
  - There are 2 alternatives here, list all snaps for *ImageName* and check for existence OR attempt creation of *snap-ImageName* and detect and work with pre-existing errors that may be returned
- If not found, create and get required response parameters (size, creation time, snapshot-volume_id), respond with data
- If found, query and get similar data and generate a response

DeleteVolume would be similar to the above.

> 
> can we make use of setting user metadata on snapshot about parent volume and snapshot creation time.

I think this is not required, Ceph maintains the source volume relationship as far as I can [gather](http://docs.ceph.com/docs/nautilus/rbd/rbd-snapshot/), and hence there should be no need to save extra metadata on the snapshot itself.

*creation_time* seems to a response to just the CreateSnapshot, on a subsequent CreateSnapshot, to return with the older *creation_time*, needs investigation into metadata that Ceph already maintains on the snapshot, OR, add it to *snap-ImageName* for future unpacking and response. It is assumed that *snap-name* like *volume-name* is unique per snapshot that the CO wants generated.
**NOTE:** If added to *snap-ImageName* then listing snapshots, to compare other ID names for Idempotent requirements would be mandatory, and we cannot just attempt creation and detect duplicate names, as *creation_time* would be unknown.

ShyamsundarR replied:
Implementation and omap usage uses the required CLIs to check volume presence and then act on the request as required.

gman0 commented:
What are the consequences for NS? What if the image/volume is not provisioned by CS but rather from outside (OpenStack Cinder/Manila)?

ShyamsundarR commented:
> What are the consequences for NS? What if the image/volume is not provisioned by CS but rather from outside (OpenStack Cinder/Manila)?

The NS recieves information either from volume_context and/or publish_context. Secrets were from the storageclass and now would shift to reading the same based on the volume_id and hence respective ceph Cluster.

For non-CS volumes, I assume the PV (in k8s terms) would contain enough information to pass along to the NS, we need the NS verbs to attempt getting information from the passed in parameters and hence detect the cluster and secrets, or fall back to assuming that this is a CS created volume and hence act on it as mentioned.

Can you help me with an annotation of a PV or an example of non-CS generated storage end point that reaches the NS? There are parts of the picture that I am extrapolating and want to be sure of.

gman0 commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+- Addition/Removal of Ceph cluster configuration, to the CSI plugin, requires
+  that the pod specification be updated, and thus the CS and NS pods require
+  to be restarted, in order to pick up the latest cluster details
+  - If done right, the NS restarts should not impact existing volumes that
```
Volumes mounted with ceph-fuse will have to be remounted...but there's little we can do about it ATM

ShyamsundarR replied:
There is no restart required now, as the config file can and is passed in as a config map volume to the pod, thus there are no restarts.

gman0 commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+  namespace: default
+data:
+  # ID of the admin/user
+  subjectid: BASE64-ENCODED-ID
```
I believe I already saw a conversation about the correct terminology for ceph credentials...can't remember where exactly. Anyway, `subjectid` and `credentials` seem a bit confusing. Credentials usually refer to a username-password pair - as opposed to just the "password" part. AFAIK passwords are referred to as keys in ceph. As for the username part, according to the [documentation](http://docs.ceph.com/docs/master/rados/operations/user-management/), and the way the names are constructed (`TYPE.ID` - in our case the `TYPE` will always be `client`), how about using `clientid`?

`clientid` in the role of username, `key` for password. Thoughts?

ShyamsundarR replied:
This has been renamed as adminid/user/id and adminkey/userkey as per other comments that note the same, see [here](https://github.com/ceph/ceph-csi/pull/241/files#diff-7551e1d8a03243f545e81cfd4ecd9156R25). Let me know if this works.

gman0 commented:
A question about the usage of FSID: are there any technical requirements for this to be an actual FSID as opposed to just _some_ identifier of a particular ceph cluster? IMHO in some situations a more user-friendly name would be more apt.

ShyamsundarR commented:
> A question about the usage of FSID: are there any technical requirements for this to be an actual FSID as opposed to just _some_ identifier of a particular ceph cluster? IMHO in some situations a more user-friendly name would be more apt.

The fsid can be any string that identifies the secret (in k8s) OR the directory name, to query for a particular Ceph clusters details. Further, it is used to encode the same into the VolID. It is not used to validate if said Ceph cluster has the same fsid etc.

Hence, it can be an arbitrary name, unique to each Ceph cluster that a CO uses, length restricted and immutable, as it is encoded into the VolID. The fsid meets these requirements nicely, hence is chosen rather than having the user to come with a name, but from a user friendliness POV it can be named with the above constraints.

I would have to rework the fsid encoding into the VolID, as a length preceded string, to accommodate variable name length strings, but otherwise nothing else changes.

If verbiage needs to change to state, `clusterid` can be a unique name with above restrictions, or for example can be the fsid in the spec (and resulting PRs), I can do the same. Let me know.

gman0 commented:
@ShyamsundarR in that case please do, also in your PR (and others)

ShyamsundarR commented:
> @ShyamsundarR in that case please do, also in your PR (and others)

Done in the code [PR](https://github.com/ceph/ceph-csi/pull/241), changing this doc is pending (as are other changes to the doc as per the implementation), and will be done subsequently as other PRs land that may require further doc changes. @gman0  if you require that doc reflect the state of submitted PRs as soon as possible, let me know, so that I can redo my priorities between the the code and doc.

ShyamsundarR commented:
Updated [section](https://github.com/ceph/ceph-csi/blob/e6139baaff4eca8209b38c514d37ac49b544a567/docs/design/proposals/stateless-csi-plugin.md#using-ceph-omaps-for-idempotence) on how image names and volume_name idempotence would be maintained using Ceph omaps instead of volume_name hashes and image instance IDs and encoding the same into the image and volume IDs.

Thanks to @dillaman for course correcting and suggesting omaps instead of the older scheme.

The older scheme was unwieldy in implementation and required that an in memory cache of all image hashes and its instances be created during startup by the CSI plugin, to ensure that newer volume name hashes use different instance IDs when there are collisions.

@liewegas @rootfs @gman0 @Madhu-1 @dillaman request a re-look of the new scheme to catch any potential issues. I have a working version of the new scheme for create/delete/mount of rbd, and am working on snapshots and cloning to reach feature parity with current implementation before posting it.

gman0 commented:
@ShyamsundarR nice! Is there anything else we could offload to Ceph omaps instead of storing it in k8s ConfigMaps?

ShyamsundarR commented:
> @ShyamsundarR nice! Is there anything else we could offload to Ceph omaps instead of storing it in k8s ConfigMaps?

Currently none, that a combination of image-meta keys and rbd info <image> does not get us. BUT, if we needed to stash away per image omaps for additional data we could extend it to create such omaps for attribute storage in the future.

I am working through snapshot code changes to use this scheme, maybe that will reveal information that we want to stash in addition to what we store now. Although my first inclination would be to store these as image-meta keys than a new omap.

ShyamsundarR commented:
> @ShyamsundarR nice! Is there anything else we could offload to Ceph omaps instead of storing it in k8s ConfigMaps?
> 
> Currently none, that a combination of image-meta keys and rbd info does not get us. BUT, if we needed to stash away per image omaps for additional data we could extend it to create such omaps for attribute storage in the future.
> 
> I am working through snapshot code changes to use this scheme, maybe that will reveal information that we want to stash in addition to what we store now. Although my first inclination would be to store these as image-meta keys than a new omap.

Hmmm... looks like image-meta keys are shared across clones and cannot be set for snapshots, so I guess we will be using a per image/snap/clone omap to store the csi.volume.name and also if required other csi keys.

dillaman commented:
> Hmmm... looks like image-meta keys are shared across clones and cannot be set for snapshots, so I guess we will be using a per image/snap/clone omap to store the csi.volume.name and also if required other csi keys.

They aren't "shared" across clones, but they are copied upon creation of the clone. WRT snapshots, if each snapshot volume were to be stored as its own (cloned) image, they again would have their own image-meta.

ShyamsundarR commented:
> > Hmmm... looks like image-meta keys are shared across clones and cannot be set for snapshots, so I guess we will be using a per image/snap/clone omap to store the csi.volume.name and also if required other csi keys.
> 
> They aren't "shared" across clones, but they are copied upon creation of the clone. WRT snapshots, if each snapshot volume were to be stored as its own (cloned) image, they again would have their own image-meta.

Thanks, makes sense. Tried changing it for the cloned image and things work as expected.

For snapshots, currently we do not want to lose the parent relation, possibly even when cloned, and also, I do not want to tie the design to either possibility.

Hence, instead of the backpointer to the k8s volname omap, being an image-meta value on the image, I think a per image/snap/clone omap to store the same sounds better (I think this is what you were referring to as well in our off-list conversation about how rbd itself does ""rbd_id.<image name>" object per image" if I am not mistaken)

dillaman commented:
> For snapshots, currently we do not want to lose the parent relation, possibly even when cloned, and also, I do not want to tie the design to either possibility.

I think it's important to not let the current RBD implementation details drive the design of the CSI. As discussed in issue #227, if the CSI spec seems to prefer that snapshots be independent, shouldn't that be the goal? You would still need to maintain metadata about which snapshots are actually owned by the CO and be able to efficiently list them. i.e. you would also want to create a directory listing for snapshots to RBD images (or an attribute on the ```csi-vol-[uuid]``` object omap keys to indicate that it's a PV or snapshot).

Also, thinking ahead to the "cloning PVs from a golden image (i.e. a snapshot)", since there is an effective limit for how many parents an image can have, having the ability to flatten a snapshot upon its first clone usage might be a nice optimization to avoid potentially needing to flatten every derived PV clone.

ShyamsundarR commented:
> > For snapshots, currently we do not want to lose the parent relation, possibly even when cloned, and also, I do not want to tie the design to either possibility.
> 
> I think it's important to not let the current RBD implementation details drive the design of the CSI. As discussed in issue #227, if the CSI spec seems to prefer that snapshots be independent, shouldn't that be the goal?

Agree, I have not yet chased the snapshot to clone and CSI/RBD perspective, to make further observations at this point in time.

> You would still need to maintain metadata about which snapshots are actually owned by the CO and be able to efficiently list them. i.e. you would also want to create a directory listing for snapshots to RBD images (or an attribute on the `csi-vol-[uuid]` object omap keys to indicate that it's a PV or snapshot).

I agree the metadata around snapshots is required, and my [comment](https://github.com/ceph/ceph-csi/pull/224#issuecomment-473959264) is ill formed and stated. What I really meant was, I would not attempt and pursue image-meta based tags (even though these are copied to a clone, and can be set to different values), but rather use a per-image omap instead, to store additional attributes as required.

Also, as stated above, snapshot metadata would also need to be preserved.

> 
> Also, thinking ahead to the "cloning PVs from a golden image (i.e. a snapshot)", since there is an effective limit for how many parents an image can have, having the ability to flatten a snapshot upon its first clone usage might be a nice optimization to avoid potentially needing to flatten every derived PV clone.

dillaman commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,511 @@
+
+ImageName format: `csi-vol-[uuid]`
+
+volume_id format: `[version]-[cluster-id]-[pool-id]-[ImageName]`
```
Probably a stupid question, but if the ```cluster-id``` is used to look-up the mon addresses, IDs, and keys, would it not be possible just have each RBD ```StorageClass``` use its own "Secret" to encode the pool id as well (i.e. the ```cluster-id``` becomes the ```storage-class-id```)?

dillaman replied:
... probably would help it be more consistent w/ CephFS as well since instead of a ```cluster-id``` and ```fscid``` to identify the CephFS "volume" (instead of ```pool-id```), it too could reference a unique persistent config structure.

ShyamsundarR replied:
> Probably a stupid question, but if the `cluster-id` is used to look-up the mon addresses, IDs, and keys, would it not be possible just have each RBD `StorageClass` use its own "Secret" to encode the pool id as well (i.e. the `cluster-id` becomes the `storage-class-id`)?

The intention is to keep a cluster level configuration entity, that can be reused across pools for the same cluster. So, a single cluster ID to denote cluster level configuration, and the StorageClass carries the pool name.

If it is required that per pool user/adminID and keys be used, then these are can be skipped in the cluster configuration secret, but passed as separate secret and passed in via the csi.storage.k8s.io/*-secret-[name|namespace] options in the StorageClass.

dillaman replied:
... and you cannot just pull the ```StorageClass``` config in the same way it is pulling the ```Secret``` (or pull a second ```Secret``` with ```StorageClass```-unique fields)? Just strikes me as odd that you need to encode the pool ID into the volume/snapshot ID but still pull the remainder of connection details from a config file.

ShyamsundarR replied:
[Secrets](https://github.com/container-storage-interface/spec/blob/c3707a60a2d4a0a684ba342c048534defb74853d/lib/go/csi/csi.pb.go#L821) are sent in separately via the RPC call, and the [options](https://github.com/container-storage-interface/spec/blob/c3707a60a2d4a0a684ba342c048534defb74853d/lib/go/csi/csi.pb.go#L817) are sent in differently.

Post volume creation, certain calls only send in the [secret](https://github.com/container-storage-interface/spec/blob/c3707a60a2d4a0a684ba342c048534defb74853d/lib/go/csi/csi.pb.go#L1895) and not the options, and hence the pool has to be encoded in the VolumeID that the plugin uses to identify the volume.

Does this explain the observation?

dillaman replied:
The secrets are sent or they are queried from k8s_store?

ShyamsundarR replied:
The secrets in the keys named `csi.storage.k8s.io/*-secret-[name|namespace]` in the storage class are sent, the [secret](https://github.com/ceph/ceph-csi/blob/csi-v1.0/examples/rbd/template-ceph-cluster-ID-secret.yaml) detailing the configuration is passed to the CSI plugin, hence queried/read from k8s_store.

dillaman replied:
OK, so it's basically a k8s built-in to send a single named secret to the controller/node agents. And because you don't want to force a 1 StorageClass-to-1 Secret mapping, the pool ID needs to get encoded in the volume/snapshot ID. However, if you did force a 1-to-1 mapping, there would be no need (similar argument for CephFS file system ID replacing pool id).

As an aside, that current "pools" parameter in the secret doesn't seem to make much sense.

ShyamsundarR replied:
Rehashing this conversation so that I am not missing the context of the discussion:

The comment is that, we could/should move the `pool` name key out of the StorageClass and add it to the secrets referred to by a given `ClusterID`, thus removing it from being encoded in the returned `VolumeID`.

### Why the current scheme:

The current scheme is (as mentioned above) to retain a 1:1 Ceph-cluster:ClusterID configuration mapping, such that when MONs or credentials of the Ceph cluster changes, it needs to be updated in one place.

The `pools` parameter in the `clusterID` named secret, is retained for listing operations. During listing, we cannot know which pools were involved in doling out PVCs and hence which ones to list from (unless we decide to query all pools to see if the omap key is present).

### What if we move the scheme to a pool:clusterID mapping:

If we had a pool:clusterID mapping, then changes to cluster configuration needs to be edited in multiple `ClusterID` named secrets.

OTOH, listing does not need `pools` key in the `ClusterID` configuration. Also, pool need not be decoded from the VolumeID and can be referred to from the `ClusterID` configuration.

Either scheme has its pros and cons as I see it.

Thoughts?

ShyamsundarR replied:
Pool name is retained in the storage class (and fsname for CephFS plugin will be added), as per various reviews and discussions across the implementation.

ShyamsundarR commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+cluster that is fully defined by the StorageClass, but brings in the burden of
+storing and managing provisioned volumes by the plugin.
+
+The proposal laid out here is to try and retain the ability for a single CSI
```
Removed multiple controller service possibility from the text.

ShyamsundarR commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,500 @@
+
+### Summary
+
+- Create a config map per Ceph cluster called `ceph-cluster-<cluster-fsid>` and
```
Secrets are now passed as per kubernetes specifications, hence there is no additional secret that is passed to the plugin.

ShyamsundarR commented:
Updated the design for CephFS related changes. The changes in short are,

- Use fsname in the CephFS StorageClass to direct provisioning to a specific fsname instance
- Use the fsname-metadata pool to store the omaps that retain volume name to image name and volume ID representation
- Store the csi objects and omaps in a RADOS namespace unique to CSI, to safeguard from other objects in the fsname-metadata pool
- The VolumeID would be encoded with the fscid of the fsname for CephFS, same as the poolID in case of RBD

The omap scheme remains the same as implemented for RBD.

Sharding the omap remains a future concern, with the ability to shard the objects into buckets based on the CSI generated volume name.

NOTE: Using `ceph volume` based commands to manage subvolumes creation and deletion would be done post the current implementation.

@batrick @ajarr Please check if the above addresses concerns around implementing the same scheme with CephFS, as per conversations over mail. If so please drop a comment, such that we can unfreeze the RBD implementation (#312 ) from a "do not merge" state to review pending state, as it is waiting on design acceptance of CephFS changes, as one of the criteria to be merged. (tagging @dillaman as well for comments provided).

ajarr commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,507 @@
+  # [Mandatory] carries the pool name to use in the Ceph cluster defined
+  #     by clusterID
+  pool: <pool-name>
+  # [Mandatory] All of the following are optional, and as described in,
```
All of the following are optional or mandatory?

ShyamsundarR replied:
The following are mandatory, need to change the text in both places.

ajarr commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,507 @@
+#### Name terminology
+
+- **ImageName**: Name of the image as created on Ceph (rbd image name, or
+  CephFS sub-directory)
```
sub-directory name?

ShyamsundarR replied:
Yes, using the same keyword ImageName to denote a RBD image name or the name of a CephFS sub-directory. Is there something missing?

ajarr commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,507 @@
+#### Using Ceph OMaps for idempotent RPCs
+
+To preserve the idempotent CreateVolume RPC requirement, across *ImageName* for
+a given *volume_name*, it is proposed to use a per-CSI instance Ceph omap named
```
you mean Ceph object named `csi.volumes.[csi-id]` with omap keys `csi.volumes.[vol_name]` and values 'imagenames'

ShyamsundarR replied:
If those are the right terms then yes, will modify the doc to word it so.

ajarr commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,507 @@
+Due to the existence of the omap, we choose to create an ImageName that
+is independent of any implicit *volume_name* or *csi-id* encoding. The image
+once created would carry a `image-meta` key:value with, the key named
+`csi.volume.name` and the value containing the CO generated *volume_name*. This
```
I'm confused. the value is the the ImageName, right? So the CO generated *volume_name* is same as the *ImageName*?

ShyamsundarR replied:
This needs to be removed, it is a carry over from past iterations. There are no image meta keys stored on the image any further (and hence none stored on CephFS sub-directory as well)

ajarr commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,507 @@
+
+- **ImageName**: Name of the image as created on Ceph (rbd image name, or
+  CephFS sub-directory)
+- **volume_id**: Identifier for the *ImageName* as shared with the CO system
```
what generates this?  this doesn't need to be stored persistently, but will be generated each time, right?

ShyamsundarR replied:
The CSI plugin generates this, and its format is detailed below.

It is generated for each volume that is provisioned using CSI, and later is used to detect which volume is being referenced based on the volume_id being passed back. So it is used to find the volume being referred to, and not generated each time.

ajarr commented on docs/design/proposals/stateless-csi-plugin.md:
```diff
@@ -0,0 +1,507 @@
+
+ImageName format: `csi-vol-[uuid]`
+
+volume_id format: `[version]-[cluster-id]-[pool-id|fscid]-[uuid]`
```
[version], version of what?

ShyamsundarR replied:
Version of the volume-id encoding format, helps adapt to future changes to the encoding format (for example adding a namespace ID when supporting using pool namespaces to create RBD images)

Madhu-1 commented:
we can update the design doc if something needed or we have enhanced the design may be in the followup PR

humblec commented:
@ShyamsundarR is this proposal matches with the current code or implementation in place? just making sure both are in parity. If you can confirm we can get this merged.
